clash-prelude/src/Clash/XException/MaybeX.hs

```diff
@@ -0,0 +1,132 @@
+  ( MaybeX(IsX, IsDefined)
+
+    -- * Construction
+  , toMaybeX
+  , hasXToMaybeX
+
+    -- * Deconstruction
+  , unsafeToA
+
+    -- * Operations
+  , andX
+  , orX
+  , maybeX
+  ) where
+
+import Prelude
+
+import Control.Applicative
+import Control.DeepSeq (NFData)
+import Control.Exception (throw)
+
+import Clash.XException (XException(..), isX, hasX)
+
+-- | Structure helping programmers to deal with 'Clash.XException.XException'
+-- values. For safety reasons it can't be constructed directly, but should be
+-- constructed using either 'pure' or 'toMaybeX'. After construction, it can be
+-- deconstructed using either 'IsX' or 'IsDefined'.
+data MaybeX a
+  = IsX_ String
+  -- ^ Upon construction, @a@ evaluated to 'Clash.XException.XException'
+  | IsDefined_ !a
+  -- ^ Upon construction, @a@ evaluated to a non-bottom WHNF
+
+instance Show a => Show (MaybeX a) where
+  showsPrec d = \case
+    IsX_ msg     -> showParen (d > 10) $ showString "IsX "       . showsPrec 11 msg
+    IsDefined_ a -> showParen (d > 10) $ showString "IsDefined " . showsPrec 11 a
+
+-- | Upon construction, @a@ evaluated to 'Clash.XException.XException'
+pattern IsX :: forall a. String -> MaybeX a
+pattern IsX msg <- IsX_ msg
+
+-- | Upon construction, @a@ evaluated to a non-bottom WHNF
+pattern IsDefined :: forall a. a -> MaybeX a
+pattern IsDefined a <- IsDefined_ a
+{-# COMPLETE IsX, IsDefined #-}
+
+instance Functor MaybeX where
+  fmap _f (IsX_ msg) = IsX_ msg
+  fmap f  (IsDefined_ a) = either IsX_ IsDefined_ (isX (f a))
+
+instance Applicative MaybeX where
+  pure = either IsX_ IsDefined_ . isX
+
+  liftA2 f (IsDefined_ a) (IsDefined_ b) = IsDefined_ (f a b)
+  liftA2 _ (IsX_ msg)     _              = IsX_ msg
+  liftA2 _ _              (IsX_ msg)     = IsX_ msg
+
+-- | Construct a 'MaybeX' value. If @a@ evaluates to 'Clash.XException.XException',
+-- this function will return 'IsX'. Otherwise, it will return 'IsDefined'.
+toMaybeX :: a -> MaybeX a
+toMaybeX = pure
+
+-- | Construct a 'MaybeX' value. If @force a@ evaluates to
+-- 'Clash.XException.XException', this function will return 'IsX'. Otherwise,
+-- it will return 'IsDefined'.
+hasXToMaybeX :: NFData a => a -> MaybeX a
+hasXToMaybeX = either IsX_ IsDefined_ . hasX
+
+-- | Deconstruct 'MaybeX' into an @a@ - the opposite of 'toMaybeX'. Note that this
+-- is unsafe, because it might return an 'Clash.XException.XException' if the
+-- argument was 'IsX'. Use of this function should be accompanied with a comment
+-- explaining why it is fine to use.
+unsafeToA :: MaybeX a -> a
+unsafeToA = maybeX (throw . XException) id
+
+-- | Map functions over both constructors.
+maybeX :: (String -> b) -> (a -> b) -> MaybeX a -> b
+maybeX f _ (IsX_ msg)     = f msg
+maybeX _ g (IsDefined_ a) = g a
+
+-- | Implements '&&' accounting for X
+--
+-- @
+-- |   | X | 1 | 0 |
+-- |---|---|---|---|
+-- | X | X | X | 0 |
+-- | 1 | X | 1 | 0 |
+-- | 0 | 0 | 0 | 0 |
+-- @
+--
+andX :: MaybeX Bool -> MaybeX Bool -> MaybeX Bool
```

Comment:
As discussed, an unreadable version of the table that renders nicely into Haddock (the character references are em spaces, widening the table cells and center aligning). If you don't like the non-Haddock version, feel free to rephrase or drop that.
```suggestion
-- | Implements '&&' accounting for X
--
-- +-------------------------+-------------------------+-------------------------+-------------------------+
-- |                         | &#x2003;__@X@__&#x2003; | &#x2003;__@1@__&#x2003; | &#x2003;__@0@__&#x2003; |
-- +-------------------------+-------------------------+-------------------------+-------------------------+
-- | &#x2003;__@X@__&#x2003; | &#x2003;@X@&#x2003;     | &#x2003;@X@&#x2003;     | &#x2003;@0@&#x2003;     |
-- +-------------------------+-------------------------+-------------------------+-------------------------+
-- | &#x2003;__@1@__&#x2003; | &#x2003;@X@&#x2003;     | &#x2003;@1@&#x2003;     | &#x2003;@0@&#x2003;     |
-- +-------------------------+-------------------------+-------------------------+-------------------------+
-- | &#x2003;__@0@__&#x2003; | &#x2003;@0@&#x2003;     | &#x2003;@0@&#x2003;     | &#x2003;@0@&#x2003;     |
-- +-------------------------+-------------------------+-------------------------+-------------------------+

-- (This is not part of the Haddock, a more readable version of the table
-- above)
--    | X | 1 | 0
-- ---|---|---|---
--  X | X | X | 0
--  1 | X | 1 | 0
--  0 | 0 | 0 | 0
andX :: MaybeX Bool -> MaybeX Bool -> MaybeX Bool
```